youtube-video-element.js

```diff
@@ -0,0 +1,542 @@
+    -1 (unstarted)
+    0 (ended)
+    1 (playing)
+    2 (paused)
+    3 (buffering)
+    5 (video cued).
+  */
+
+  get paused() {
+    if (!this.isLoaded) return !this.autoplay;
+    return [-1, 0, 2, 5].includes(this.api?.getPlayerState?.());
+  }
+
+  get duration() {
+    return this.api?.getDuration?.() ?? NaN;
+  }
+
+  get autoplay() {
+    return this.hasAttribute('autoplay');
+  }
+
+  set autoplay(val) {
+    if (this.autoplay == val) return;
+    if (val) this.setAttribute('autoplay', '');
+    else this.removeAttribute('autoplay');
+  }
+
+  get buffered() {
+    if (!this.isLoaded) return createTimeRanges();
+    const progress =
+      this.api?.getVideoLoadedFraction() * this.api?.getDuration();
+    if (progress > 0) {
+      return createTimeRanges(0, progress);
+    }
+    return createTimeRanges();
+  }
+
+  get controls() {
+    return this.hasAttribute('controls');
+  }
+
+  set controls(val) {
+    if (this.controls == val) return;
+    if (val) this.setAttribute('controls', '');
+    else this.removeAttribute('controls');
+  }
+
+  get currentTime() {
+    return this.api?.getCurrentTime?.() ?? 0;
+  }
+
+  set currentTime(val) {
+    if (this.currentTime == val) return;
+    this.loadComplete.then(() => {
+      this.api?.seekTo(val, true);
+      if (this.paused) {
+        this.pause();
+      }
+    });
+  }
+
+  set defaultMuted(val) {
+    if (this.defaultMuted == val) return;
+    if (val) this.setAttribute('muted', '');
+    else this.removeAttribute('muted');
+  }
+
+  get defaultMuted() {
+    return this.hasAttribute('muted');
+  }
+
+  get loop() {
+    return this.hasAttribute('loop');
+  }
+
+  set loop(val) {
+    if (this.loop == val) return;
+    if (val) this.setAttribute('loop', '');
+    else this.removeAttribute('loop');
+  }
+
+  set muted(val) {
+    if (this.muted == val) return;
+    this.loadComplete.then(() => {
+      val ? this.api?.mute() : this.api?.unMute();
+    });
+  }
+
+  get muted() {
+    if (!this.isLoaded) return this.defaultMuted;
+    return this.api?.isMuted?.();
+  }
+
+  get playbackRate() {
+    return this.api?.getPlaybackRate?.() ?? 1;
+  }
+
+  set playbackRate(val) {
+    if (this.playbackRate == val) return;
+    this.loadComplete.then(() => {
+      this.api?.setPlaybackRate(val);
+    });
+  }
+
+  get playsInline() {
+    return this.hasAttribute('playsinline');
+  }
+
+  set playsInline(val) {
+    if (this.playsInline == val) return;
+    if (val) this.setAttribute('playsinline', '');
+    else this.removeAttribute('playsinline');
+  }
+
+  get poster() {
+    return this.getAttribute('poster');
+  }
+
+  set poster(val) {
+    if (this.poster == val) return;
+    this.setAttribute('poster', `${val}`);
+  }
+
+  set volume(val) {
+    if (this.volume == val) return;
+    this.loadComplete.then(() => {
+      this.api?.setVolume(val * 100);
+    });
+  }
+
+  get volume() {
+    if (!this.isLoaded) return 1;
+    return this.api?.getVolume() / 100;
+  }
```

Comment:
I wonder if we need to keep a cache of values like for volume so that you can get the updated value immediately, since the operation is async. Probably not a blocker, but something to think about.